### PR TITLE
Fix Crash When Enabling and Disabling ETW with Old Callbacks

### PR DESCRIPTION
**Pull request by ivberg**
### Description
Under certain conditions with enabling & disabling ETW continuously, we got a crash report.
Allows ETW callbacks to be de-registered upon class destructor.
Related to #20537

### Motivation and Context
Fixes crash

### Callstack
We see it crash in
[0x0]   onnxruntime!<lambda_967a738fca8512372f170fcaf2d094d4>::operator()+0x34   0x12941ff570   0x7ffa994f0a04   

[0x1]   onnxruntime!std::_Func_class<void,_GUID const *,unsigned long,unsigned char,unsigned __int64,unsigned __int64,_EVENT_FILTER_DESCRIPTOR *,void *>::operator()+0x54   0x12941ff7b0   0x7ffa994f0d64   

[0x2]   onnxruntime!onnxruntime::logging::EtwRegistrationManager::InvokeCallbacks+0xcc   0x12941ff7b0   0x7ffa994f0d64   

[0x3]   onnxruntime!onnxruntime::logging::EtwRegistrationManager::ORT_TL_EtwEnableCallback+0x94   0x12941ff860   0x7ffa98d19628   
 

and seems to us that the this pointer captured in 
etwRegistrationManager.RegisterInternalCallback(
      [&etwRegistrationManager, this](
...
is no longer valid when the callback is called.

### CODE REVIEW

File: onnxruntime/core/framework/execution_providers.h [hunk 1]
```diff
@@ -24,31 +24,15 @@ namespace onnxruntime {
...
-
-    ORT_IGNORE_RETURN_VALUE(provider_idx_map_.insert({provider_id, new_provider_idx}));
-
-    // update execution provider options
-    auto providerOptions = p_exec_provider->GetProviderOptions();
-    exec_provider_options_[provider_id] = providerOptions;
-
 #ifdef _WIN32
-    LogProviderOptions(provider_id, providerOptions, false);
+  WindowsTelemetry::EtwInternalCallback etw_callback_;
```
Comment (skottmckay) [inline onnxruntime/core/framework/execution_providers.h:28]: nit: can we move this to the end of the class with all the other member declarations?

File: onnxruntime/core/framework/execution_providers.h [hunk 2]
```diff
@@ -156,5 +170,5 @@ class ExecutionProviders {
   // Whether the CPU provider was implicitly added to a session for fallback (true),
   // or whether it was explicitly added by the caller.
   bool cpu_execution_provider_was_implicitly_added_ = false;
-};
+};  // namespace onnxruntime
```
Comment (skottmckay) [inline onnxruntime/core/framework/execution_providers.h:173]:
```suggestion
};
```

File: onnxruntime/core/platform/windows/logging/etw_sink.h [hunk 3]
```diff
@@ -71,6 +71,8 @@ class EtwRegistrationManager {
 
   void RegisterInternalCallback(const EtwInternalCallback& callback);
 
+  void UnregisterInternalCallback(const EtwInternalCallback& callback);
```
Comment (skottmckay) [inline onnxruntime/core/platform/windows/logging/etw_sink.h:74]: nit: looks like there's duplicated functionality for handling the callbacks (Register/Invoke/Unregister) that could maybe be split out into a helper class that WindowsTelemetry and EtwSink use. not needed for this PR but something to keep in mind next time the setup is being modified.

File: onnxruntime/core/framework/execution_providers.h (hunk 2 above)
Comment (ivberg) [reply onnxruntime/core/framework/execution_providers.h:173]: Don't know how that snuck in there. Fixed

File: onnxruntime/core/platform/windows/logging/etw_sink.h (hunk 3 above)
Comment (ivberg) [reply onnxruntime/core/platform/windows/logging/etw_sink.h:74]: yea, next time

**Review by skottmckay — APPROVED** [review]

**Pull request merged**